src/traces/volume/attributes.js

```diff
@@ -0,0 +1,127 @@
+/**
+* Copyright 2012-2018, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var colorscaleAttrs = require('../../components/colorscale/attributes');
+var colorbarAttrs = require('../../components/colorbar/attributes');
+var mesh3dAttrs = require('../mesh3d/attributes');
+var baseAttrs = require('../../plots/attributes');
+
+var extendFlat = require('../../lib/extend').extendFlat;
+
+var attrs = {
+    x: {
+        valType: 'data_array',
+        role: 'info',
+        editType: 'calc+clearAxisTypes',
+        description: [
+            'Sets the x coordinates of the volume'
+        ].join(' ')
+    },
+    y: {
+        valType: 'data_array',
+        role: 'info',
+        editType: 'calc+clearAxisTypes',
+        description: [
+            'Sets the y coordinates of the volume'
+        ].join(' ')
+    },
+    z: {
+        valType: 'data_array',
+        role: 'info',
+        editType: 'calc+clearAxisTypes',
+        description: [
+            'Sets the z coordinates of the volume'
+        ].join(' ')
+    },
+
+    value: {
+        valType: 'data_array',
+        role: 'info',
+        editType: 'calc',
+        description: 'Sets the intensity values of the volume.'
+    },
+
+    opacityscale: {
+        valType: 'data_array',
+        role: 'info',
+        editType: 'calc',
+        description: 'Sets the opacity scale of the volume, which opacity to use for which intensity. Array of 256 values in 0..1 range.'
+    },
+
+    imin: {
+        valType: 'number',
+        role: 'info',
+        editType: 'calc',
+        description: 'Sets the minimum intensity bound of the volume.'
+    },
+
+    imax: {
+        valType: 'number',
+        role: 'info',
+        editType: 'calc',
+        description: 'Sets the maximum intensity bound of the volume.'
+    },
+
+    opacity: {
+        valType: 'number',
+        role: 'info',
+        editType: 'calc',
+        description: 'Sets the opacity of the volume.'
+    },
```

Comment:
`opacity` is a standard attribute for all traces (unless you disable it with `'noOpacity'` in `module.categories`) so you shouldn't need this attribute, nor the `coerce('opacity')`, nor even `trace.opacity === undefined ? 1 : trace.opacity` - you should just be able to do `volumeOpts.opacity = trace.opacity;` since opacity defaults to 1.